File: earthsciio/backends/cds.py
"""The active ``cds`` transport — Copernicus Climate Data Store API v1.

Registered under the URL scheme ``cds`` (``spec/registries.md`` §1). Unlike a
plain ``http`` GET, a CDS fetch is a **three-step async retrieval**: submit a
request, poll the job until it succeeds, then download the produced asset. This
module ports the Julia ``cds_api.jl`` client to plain ``requests`` (no Python
``cdsapi`` SDK dependency) and wraps it behind the standard
:class:`~earthsciio.registry.Transport` interface so the content-addressed cache
dispatches it exactly like every other transport.

**The ``cds://`` URL.** The transport interface receives a single resolved URL,
but a CDS retrieval needs a *dataset* plus a *request Dict*. Both are encoded
into the URL by :func:`encode_cds_url` — the cross-language form
``cds://<dataset>?<canonical-request-json>`` (``spec/registries.md`` §1) — and
recovered by :func:`decode_cds_url`. The JSON is serialized with recursively
sorted keys and compact separators, byte-identical to the Julia/Rust tracks, so
an identical request always yields an identical URL — and therefore an identical
cache key (``sha256(resolved_url)``) **across languages**. That is what makes the
cache's **skip-if-exists** work: re-requesting the same ERA5 month is a cache
hit, never a second CDS job. The loader-side request builders live in
:mod:`earthsciio.era5`.

**Authentication.** The CDS API authenticates with a ``PRIVATE-TOKEN`` header.
The spec-aligned path injects it through the pluggable auth seam: register
:func:`cds_auth` under the ``cds`` realm on the :class:`~earthsciio.cache.Cache`
and fetch with ``auth_realm="cds"`` — credentials reach the request, never the
manifest. For standalone use (and the live smoke test) the transport falls back
to reading the key itself from ``$CDSAPI_KEY`` or ``~/.cdsapirc`` via
:func:`cds_api_key`, mirroring the Julia client.

**Offline mode.** Like every transport, ``cds`` is never even constructed when
``offline=True`` — the cache serves the previously downloaded blob from the
store and never touches the network.
"""

from __future__ import annotations

import json
import os
import pathlib
import time
from typing import Any, Dict, List, Mapping, Optional, Tuple

from ..auth import StaticHeaderAuth
from ..errors import TransportError
from ..transport import DOWNLOADED, FetchResult

__all__ = [
    "CdsTransport",
    "CDS_SCHEME",
    "DEFAULT_CDS_API_URL",
    "cds_api_key",
    "cds_api_url",
    "cds_auth",
    "encode_cds_url",
    "decode_cds_url",
    "cds_submit",
    "cds_wait",
    "cds_download",
]

#: The URL scheme this transport answers to (its ``transport`` registry key).
CDS_SCHEME = "cds"

#: The CDS API v1 root. The NEW endpoint (the legacy ``/v2`` host is retired).
DEFAULT_CDS_API_URL = "https://cds.climate.copernicus.eu/api"

#: Seconds between job-status polls, and the overall give-up budget (ported from
#: ``cds_api.jl``'s ``CDS_POLL_INTERVAL`` / ``CDS_TIMEOUT``).
DEFAULT_POLL_INTERVAL = 5.0
DEFAULT_TIMEOUT = 600.0

#: The auth realm name recorded in the manifest for a CDS fetch (never the key).
CDS_REALM = "cds"

#: The CDS header that carries the API key.
_TOKEN_HEADER = "PRIVATE-TOKEN"

_USER_AGENT = "earthsciio/python"
_CHUNK = 1 << 16
#: (connect, read) timeouts in seconds — a wedged CDS endpoint must not hang.
_HTTP_TIMEOUT = (10, 300)


# --------------------------------------------------------------------------- #
# Credentials + endpoint resolution (port of cds_api.jl `cds_api_key`).
# --------------------------------------------------------------------------- #


def _cdsapirc_path() -> pathlib.Path:
    return pathlib.Path(os.path.expanduser("~")) / ".cdsapirc"


def _read_cdsapirc() -> Dict[str, str]:
    """Parse ``~/.cdsapirc`` into ``{"url": ..., "key": ...}`` (missing keys absent).

    The file format is the two-line ``url: ...`` / ``key: ...`` used by the CDS
    SDK. Whitespace around the value is stripped; lines without a recognized
    ``name:`` prefix are ignored. A missing file yields an empty mapping.
    """
    rc = _cdsapirc_path()
    out: Dict[str, str] = {}
    try:
        text = rc.read_text()
    except (FileNotFoundError, OSError):
        return out
    for line in text.splitlines():
        name, sep, value = line.partition(":")
        if not sep:
            continue
        name = name.strip().lower()
        if name in ("url", "key") and name not in out:
            out[name] = value.strip()
    return out


def cds_api_key() -> str:
    """The CDS API key from ``$CDSAPI_KEY`` or the ``key:`` line of ``~/.cdsapirc``.

    Precedence matches the Julia client: the environment variable wins, then the
    rc file. Raises :class:`~earthsciio.errors.TransportError` when neither is
    present — a hard, surfaced configuration error, not a silent unauthenticated
    fetch.
    """
    env = os.environ.get("CDSAPI_KEY")
    if env:
        return env.strip()
    rc = _read_cdsapirc()
    if rc.get("key"):
        return rc["key"]
    raise TransportError(
        "CDS API key not found. Set $CDSAPI_KEY or create ~/.cdsapirc with "
        "'key: <your-key>' (see https://cds.climate.copernicus.eu/how-to-api)."
    )


def cds_api_url() -> str:
    """The CDS API root: ``$CDSAPI_URL`` → ``url:`` in ``~/.cdsapirc`` → default.

    Overridable so a fetch can target a test/mock server or a mirror without
    touching the transport. The trailing slash is normalized off so the client
    can join paths unambiguously.
    """
    env = os.environ.get("CDSAPI_URL")
    if env:
        return env.strip().rstrip("/")
    rc = _read_cdsapirc()
    if rc.get("url"):
        return rc["url"].rstrip("/")
    return DEFAULT_CDS_API_URL


def cds_auth(key: Optional[str] = None) -> StaticHeaderAuth:
    """A ``cds``-realm resolver carrying the ``PRIVATE-TOKEN`` header.

    Register it on the cache (``auth={"cds": cds_auth()}``) and fetch with
    ``auth_realm="cds"``. With ``key=None`` the token is read once via
    :func:`cds_api_key`; the caller may also pass a key resolved from its own
    secret source. The realm *name* is all that ever reaches the manifest.
    """
    token = key if key is not None else cds_api_key()
    return StaticHeaderAuth.header(CDS_REALM, _TOKEN_HEADER, token)


# --------------------------------------------------------------------------- #
# The cds:// URL codec — (dataset, request) <-> a resolved, cache-keyable URL.
# --------------------------------------------------------------------------- #


def _canonical_request_json(request: Mapping[str, Any]) -> str:
    """Deterministic JSON for a CDS request: recursively sorted keys, no spaces.

    Byte-identical to the Julia/Rust tracks' canonical encoders (Rust's
    ``serde_json`` over a ``BTreeMap``, Julia's ``_canonical_json``), so the same
    logical request hashes to the same cache key in every language.
    """
    return json.dumps(request, sort_keys=True, separators=(",", ":"))


def encode_cds_url(dataset: str, request: Mapping[str, Any]) -> str:
    """Encode a CDS ``(dataset, request)`` pair into a resolved ``cds://`` URL.

    The shared, cross-language form is ``cds://<dataset>?<canonical-request-json>``
    (``spec/registries.md`` §1): the request is appended as canonical JSON
    verbatim — **no** ``request=`` parameter and **no** percent-encoding — so the
    URL string (and therefore ``sha256(resolved_url)``) is byte-identical to the
    Rust track for the same request, and a repeat is a cross-language cache hit.
    The dataset is the authority and must not contain ``?``.
    """
    if not dataset:
        raise ValueError("CDS dataset id must be non-empty")
    if "?" in dataset:
        raise ValueError(f"CDS dataset id must not contain '?': {dataset!r}")
    return f"{CDS_SCHEME}://{dataset}?{_canonical_request_json(request)}"


def decode_cds_url(url: str) -> Tuple[str, Dict[str, Any]]:
    """Recover ``(dataset, request)`` from a ``cds://`` URL (inverse of encode).

    String-split (not ``urlsplit``) so the raw JSON in the query is preserved
    verbatim, mirroring the Rust ``parse_cds_url``. Raises :class:`ValueError`
    for a non-``cds`` URL, a missing/empty request, or a request that is not a
    JSON object — a transport must never guess at a corrupt resolved URL.
    """
    prefix = f"{CDS_SCHEME}://"
    if not url.startswith(prefix):
        raise ValueError(f"not a cds:// URL: {url!r}")
    dataset, sep, payload = url[len(prefix):].partition("?")
    if not sep:
        raise ValueError(f"cds:// URL missing '?<request-json>': {url!r}")
    if not dataset:
        raise ValueError(f"cds:// URL has an empty dataset: {url!r}")
    if not payload:
        raise ValueError(f"cds:// URL has an empty request: {url!r}")
    try:
        request = json.loads(payload)
    except json.JSONDecodeError as exc:
        raise ValueError(f"cds:// request is not valid JSON: {url!r}") from exc
    if not isinstance(request, dict):
        raise ValueError(f"cds:// request must be a JSON object: {url!r}")
    return dataset, request


# --------------------------------------------------------------------------- #
# The low-level client (port of cds_api.jl: submit -> wait -> download).
# --------------------------------------------------------------------------- #


def cds_submit(
    api_url: str,
    dataset: str,
    request: Mapping[str, Any],
    headers: Mapping[str, str],
) -> str:
    """POST a retrieve request and return the job id (``cds_submit`` in Julia).

    Endpoint: ``{api_url}/retrieve/v1/processes/{dataset}/execution`` with body
    ``{"inputs": request}``. A queued/running/finished job all carry a ``jobID``;
    anything else is a submission failure.
    """
    import requests  # lazy: never imported on the offline path

    url = f"{api_url}/retrieve/v1/processes/{dataset}/execution"
    send = {"User-Agent": _USER_AGENT, "Content-Type": "application/json"}
    send.update(headers)
    try:
        resp = requests.post(
            url, json={"inputs": dict(request)}, headers=send, timeout=_HTTP_TIMEOUT
        )
    except requests.RequestException as exc:
        raise TransportError(f"CDS submit failed for {dataset}: {exc}") from exc
    if resp.status_code >= 400:
        raise TransportError(
            f"CDS submit returned {resp.status_code} for {dataset}: {resp.text}"
        )
    data = resp.json()
    if data.get("status") in ("accepted", "running", "successful") and data.get("jobID"):
        return str(data["jobID"])
    raise TransportError(f"CDS submit did not return a job id for {dataset}: {data!r}")


def cds_wait(
    api_url: str,
    job_id: str,
    headers: Mapping[str, str],
    *,
    poll_interval: float = DEFAULT_POLL_INTERVAL,
    timeout: float = DEFAULT_TIMEOUT,
    sleep=time.sleep,
    monotonic=time.monotonic,
) -> str:
    """Poll a job until ``successful`` and return the asset download href.

    Mirrors ``cds_wait``: on ``successful`` it GETs ``{job}/results`` and returns
    ``asset.value.href``; ``failed`` raises; exceeding ``timeout`` raises. ``sleep``
    and ``monotonic`` are injectable so tests drive the loop without real delay.
    """
    import requests  # lazy

    job_url = f"{api_url}/retrieve/v1/jobs/{job_id}"
    send = {"User-Agent": _USER_AGENT}
    send.update(headers)
    start = monotonic()
    while True:
        try:
            resp = requests.get(job_url, headers=send, timeout=_HTTP_TIMEOUT)
        except requests.RequestException as exc:
            raise TransportError(f"CDS poll failed for job {job_id}: {exc}") from exc
        if resp.status_code >= 400:
            raise TransportError(
                f"CDS poll returned {resp.status_code} for job {job_id}: {resp.text}"
            )
        status = resp.json().get("status", "")
        if status == "successful":
            try:
                results = requests.get(
                    f"{job_url}/results", headers=send, timeout=_HTTP_TIMEOUT
                )
            except requests.RequestException as exc:
                raise TransportError(
                    f"CDS results fetch failed for job {job_id}: {exc}"
                ) from exc
            if results.status_code >= 400:
                raise TransportError(
                    f"CDS results returned {results.status_code} for job {job_id}: "
                    f"{results.text}"
                )
            try:
                return results.json()["asset"]["value"]["href"]
            except (KeyError, TypeError) as exc:
                raise TransportError(
                    f"CDS results for job {job_id} missing asset href: {results.text}"
                ) from exc
        if status == "failed":
            raise TransportError(f"CDS job {job_id} failed: {resp.text}")
        if monotonic() - start > timeout:
            raise TransportError(
                f"CDS job {job_id} timed out after {timeout}s (last status: {status!r})"
            )
        sleep(poll_interval)


def cds_download(href: str, dest: Any) -> int:
    """Stream the produced asset to ``dest``; return the byte count.

    The asset href is a pre-signed object-store URL, so no auth header is sent
    (the Julia client downloads it the same way). Streams in chunks so a large
    NetCDF never lands wholly in memory.
    """
    import requests  # lazy

    try:
        resp = requests.get(
            href, stream=True, timeout=_HTTP_TIMEOUT, allow_redirects=True
        )
    except requests.RequestException as exc:
        raise TransportError(f"CDS asset download failed for {href}: {exc}") from exc
    try:
        if resp.status_code != 200:
            raise TransportError(
                f"CDS asset download returned {resp.status_code} for {href}"
            )
        written = 0
        with open(dest, "wb") as fh:
            for chunk in resp.iter_content(chunk_size=_CHUNK):
                if chunk:
                    fh.write(chunk)
                    written += len(chunk)
        return written
    finally:
        resp.close()


# --------------------------------------------------------------------------- #
# The Transport implementation.
# --------------------------------------------------------------------------- #


class CdsTransport:
    """CDS API v1 transport: submit → poll → download behind the fetch seam.

    Constructed with no arguments by the cache (``transport_registry.create("cds")``);
    the API root, poll cadence and timeout default to the module constants and
    can be overridden for tests or a mirror. ``conditional`` (ETag/If-Modified)
    has no CDS analog and is ignored — re-use is keyed on the deterministic
    ``cds://`` URL, so an unchanged request is a cache hit and this transport is
    never re-run for it.
    """

    NAME = "cds"
    SCHEMES = (CDS_SCHEME,)

    def __init__(
        self,
        api_url: Optional[str] = None,
        *,
        poll_interval: float = DEFAULT_POLL_INTERVAL,
        timeout: float = DEFAULT_TIMEOUT,
        sleep=time.sleep,
    ) -> None:
        self._api_url = api_url
        self.poll_interval = poll_interval
        self.timeout = timeout
        self._sleep = sleep

    def schemes(self) -> List[str]:
        return list(self.SCHEMES)

    def _token_headers(self, auth: Optional[Any]) -> Dict[str, str]:
        """The auth headers for the CDS API calls.

        Prefers the injected resolver (the spec-aligned path — credentials never
        baked into the transport); falls back to reading the key directly for
        standalone/live use.
        """
        if auth is not None:
            return {name: value for name, value in auth.headers()}
        return {_TOKEN_HEADER: cds_api_key()}

    def fetch(
        self,
        resolved_url: str,
        dest: Any,
        conditional: Optional[Dict[str, Any]] = None,
        auth: Optional[Any] = None,
    ) -> FetchResult:
        dataset, request = decode_cds_url(resolved_url)
        api_url = (self._api_url or cds_api_url()).rstrip("/")
        headers = self._token_headers(auth)

        job_id = cds_submit(api_url, dataset, request, headers)
        href = cds_wait(
            api_url,
            job_id,
            headers,
            poll_interval=self.poll_interval,
            timeout=self.timeout,
            sleep=self._sleep,
        )
        written = cds_download(href, dest)
        return FetchResult(DOWNLOADED, bytes_written=written)

File: rust/src/transport/cds.rs
//! The `cds` transport: the Copernicus Climate Data Store (CDS) API v1
//! (`spec/registries.md` §1). Ported from EarthSciData.jl's `cds_api.jl`.
//!
//! CDS is not a plain byte fetch. A `cds://<dataset>?<request-json>` resolved URL
//! is **submitted** (`POST .../processes/<dataset>/execution`), the returned job
//! is **polled** (`GET .../jobs/<id>`) until it succeeds, the job's **results**
//! (`GET .../jobs/<id>/results`) yield an asset `href`, and that href is finally
//! **downloaded** into the cache staging path. The whole submit→poll→download
//! dance is one [`Transport::fetch`], so the content-addressed cache treats CDS
//! exactly like any other source: same key (`sha256(resolved_url)`), same
//! skip-if-exists, same offline replay.
//!
//! Auth is the pluggable `cds` realm: a `PRIVATE-TOKEN` header carrying the key
//! from `~/.cdsapirc` or `$CDSAPI_KEY`, injected as the fetch `auth` — never
//! baked into the transport (`spec/registries.md` §1).

use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

use reqwest::blocking::Client;
use reqwest::header::{HeaderMap, HeaderName, HeaderValue, CONTENT_TYPE};

use super::{Conditional, FetchResult, FetchStatus, Transport};
use crate::auth::{AuthResolver, StaticHeaderAuth};
use crate::error::{Error, Result};

/// The production CDS API v1 endpoint.
pub const CDS_API_URL: &str = "https://cds.climate.copernicus.eu/api";

/// The auth realm CDS fetches authenticate under (the `PRIVATE-TOKEN` header).
pub const CDS_REALM: &str = "cds";

/// Default delay between job-status polls (mirrors `cds_api.jl`'s 5 s).
const DEFAULT_POLL_INTERVAL: Duration = Duration::from_secs(5);

/// Default budget before a job is abandoned (mirrors `cds_api.jl`'s 600 s).
const DEFAULT_TIMEOUT: Duration = Duration::from_secs(600);

/// The `cds://` transport: submit a CDS request, poll the job, download the
/// asset. Construct via [`CdsTransport::new`] (production endpoint) or
/// [`CdsTransport::with_base_url`] (tests / mirrors), tuning the cadence with
/// [`poll_interval`](Self::poll_interval) / [`timeout`](Self::timeout).
pub struct CdsTransport {
    client: Client,
    base_url: String,
    poll_interval: Duration,
    timeout: Duration,
}

impl CdsTransport {
    /// A transport against the production CDS endpoint with default cadence.
    pub fn new() -> Self {
        Self::with_base_url(CDS_API_URL)
    }

    /// A transport against `base_url` (a trailing slash is trimmed). Used to
    /// point at a mock server in tests, or an alternate CDS-compatible endpoint.
    pub fn with_base_url(base_url: impl Into<String>) -> Self {
        let client = Client::builder()
            .user_agent(concat!("earthsciio/", env!("CARGO_PKG_VERSION")))
            .build()
            .expect("default reqwest blocking client builds");
        Self::with_client(client, base_url)
    }

    /// A transport from a caller-provided client (custom timeouts, proxies, …).
    pub fn with_client(client: Client, base_url: impl Into<String>) -> Self {
        Self {
            client,
            base_url: base_url.into().trim_end_matches('/').to_string(),
            poll_interval: DEFAULT_POLL_INTERVAL,
            timeout: DEFAULT_TIMEOUT,
        }
    }

    /// Set the delay between job-status polls (default 5 s).
    pub fn poll_interval(mut self, d: Duration) -> Self {
        self.poll_interval = d;
        self
    }

    /// Set the budget before a still-running job is abandoned (default 600 s).
    pub fn timeout(mut self, d: Duration) -> Self {
        self.timeout = d;
        self
    }

    /// Submit a retrieve request; returns the job ID. Mirrors `cds_submit`.
    fn submit(
        &self,
        url: &str,
        dataset: &str,
        request_json: &str,
        auth: &HeaderMap,
    ) -> Result<String> {
        let exec_url = format!(
            "{}/retrieve/v1/processes/{dataset}/execution",
            self.base_url
        );
        // `request_json` is already a canonical JSON object; wrap it verbatim so
        // the body matches the cache key's request byte-for-byte.
        let body = format!("{{\"inputs\":{request_json}}}");

        let mut headers = auth.clone();
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));

        let resp = self
            .client
            .post(&exec_url)
            .headers(headers)
            .body(body)
            .send()
            .map_err(|e| transport_err(url, e.to_string()))?;
        let status = resp.status();
        let text = resp.text().map_err(|e| transport_err(url, e.to_string()))?;
        if !status.is_success() {
            return Err(transport_err(
                url,
                format!("submit HTTP {}: {text}", status.as_u16()),
            ));
        }

        let data: serde_json::Value = serde_json::from_str(&text)
            .map_err(|e| transport_err(url, format!("submit response not JSON: {e}")))?;
        match data.get("status").and_then(|v| v.as_str()).unwrap_or("") {
            "accepted" | "running" | "successful" => {}
            _ => return Err(transport_err(url, format!("CDS submit rejected: {text}"))),
        }
        let job_id = data
            .get("jobID")
            .and_then(|v| v.as_str())
            .ok_or_else(|| transport_err(url, format!("submit response missing jobID: {text}")))?;
        Ok(job_id.to_string())
    }

    /// Poll a job to completion; returns the download href. Mirrors `cds_wait`.
    fn wait(&self, url: &str, job_id: &str, auth: &HeaderMap) -> Result<String> {
        let job_url = format!("{}/retrieve/v1/jobs/{job_id}", self.base_url);
        let start = Instant::now();
        loop {
            let resp = self
                .client
                .get(&job_url)
                .headers(auth.clone())
                .send()
                .map_err(|e| transport_err(url, e.to_string()))?;
            let status = resp.status();
            let text = resp.text().map_err(|e| transport_err(url, e.to_string()))?;
            if !status.is_success() {
                return Err(transport_err(
                    url,
                    format!("poll HTTP {}: {text}", status.as_u16()),
                ));
            }
            let data: serde_json::Value = serde_json::from_str(&text)
                .map_err(|e| transport_err(url, format!("poll response not JSON: {e}")))?;

            match data.get("status").and_then(|v| v.as_str()).unwrap_or("") {
                "successful" => return self.results_href(url, &job_url, auth),
                "failed" => {
                    return Err(transport_err(
                        url,
                        format!("CDS job {job_id} failed: {text}"),
                    ))
                }
                _ => {
                    if start.elapsed() > self.timeout {
                        return Err(transport_err(
                            url,
                            format!("CDS job {job_id} timed out after {:?}", self.timeout),
                        ));
                    }
                    std::thread::sleep(self.poll_interval);
                }
            }
        }
    }

    /// Fetch the results document and extract `asset.value.href`.
    fn results_href(&self, url: &str, job_url: &str, auth: &HeaderMap) -> Result<String> {
        let results_url = format!("{job_url}/results");
        let resp = self
            .client
            .get(&results_url)
            .headers(auth.clone())
            .send()
            .map_err(|e| transport_err(url, e.to_string()))?;
        let status = resp.status();
        let text = resp.text().map_err(|e| transport_err(url, e.to_string()))?;
        if !status.is_success() {
            return Err(transport_err(
                url,
                format!("results HTTP {}: {text}", status.as_u16()),
            ));
        }
        let data: serde_json::Value = serde_json::from_str(&text)
            .map_err(|e| transport_err(url, format!("results response not JSON: {e}")))?;
        let href = data
            .pointer("/asset/value/href")
            .and_then(|v| v.as_str())
            .ok_or_else(|| {
                transport_err(url, format!("results missing asset.value.href: {text}"))
            })?;
        Ok(href.to_string())
    }

    /// Stream the asset href into the staging file; returns bytes written.
    fn download(&self, url: &str, href: &str, dest: &Path) -> Result<u64> {
        // The href is a CDS-issued (often pre-signed) URL — downloaded without
        // the PRIVATE-TOKEN, matching `cds_api.jl`'s `_download_with_progress`.
        let mut resp = self
            .client
            .get(href)
            .send()
            .map_err(|e| transport_err(url, e.to_string()))?;
        let status = resp.status();
        if !status.is_success() {
            return Err(transport_err(
                url,
                format!("download HTTP {} from {href}", status.as_u16()),
            ));
        }
        let mut file =
            std::fs::File::create(dest).map_err(|e| Error::io(Some(dest.to_path_buf()), e))?;
        let bytes_written = resp
            .copy_to(&mut file)
            .map_err(|e| transport_err(url, e.to_string()))?;
        file.flush()
            .map_err(|e| Error::io(Some(dest.to_path_buf()), e))?;
        Ok(bytes_written)
    }
}

impl Default for CdsTransport {
    fn default() -> Self {
        Self::new()
    }
}

impl Transport for CdsTransport {
    fn schemes(&self) -> &'static [&'static str] {
        &["cds"]
    }

    fn fetch(
        &self,
        url: &str,
        dest: &Path,
        _conditional: &Conditional,
        auth: Option<&dyn AuthResolver>,
    ) -> Result<FetchResult> {
        // CDS regenerates the asset on every request, so conditional GET does
        // not apply — there is nothing to revalidate against.
        let (dataset, request_json) = parse_cds_url(url)?;
        let headers = auth_headers(url, auth)?;

        let job_id = self.submit(url, &dataset, &request_json, &headers)?;
        let href = self.wait(url, &job_id, &headers)?;
        let bytes_written = self.download(url, &href, dest)?;

        Ok(FetchResult {
            status: FetchStatus::Downloaded,
            etag: None,
            last_modified: None,
            bytes_written,
        })
    }
}

/// Build a `cds://<dataset>?<request-json>` resolved URL. `request_json` must be
/// a **canonical** (deterministic) JSON encoding of the CDS request object so the
/// same logical request always yields the same cache key (`spec/cache-format.md`).
pub fn build_cds_url(dataset: &str, request_json: &str) -> String {
    format!("cds://{dataset}?{request_json}")
}

/// Split a `cds://<dataset>?<request-json>` URL into `(dataset, request_json)`.
///
/// The request is validated as a JSON object so a malformed request fails at the
/// cache boundary with a clear [`Error::BadUrl`], not as an opaque CDS 400 later.
pub fn parse_cds_url(url: &str) -> Result<(String, String)> {
    let bad = |detail: String| Error::BadUrl {
        url: url.to_string(),
        detail,
    };
    let rest = url
        .strip_prefix("cds://")
        .ok_or_else(|| bad("not a cds:// URL".to_string()))?;
    let (dataset, request_json) = rest
        .split_once('?')
        .ok_or_else(|| bad("cds:// URL missing '?<request-json>'".to_string()))?;
    if dataset.is_empty() {
        return Err(bad("cds:// URL has an empty dataset".to_string()));
    }
    if request_json.is_empty() {
        return Err(bad("cds:// URL has an empty request".to_string()));
    }
    let parsed: serde_json::Value = serde_json::from_str(request_json)
        .map_err(|e| bad(format!("cds:// request is not JSON: {e}")))?;
    if !parsed.is_object() {
        return Err(bad("cds:// request must be a JSON object".to_string()));
    }
    Ok((dataset.to_string(), request_json.to_string()))
}

/// Read the CDS API key from `$CDSAPI_KEY` or `~/.cdsapirc` (the `key:` line),
/// mirroring `cds_api.jl`'s `cds_api_key`. Errors when neither is present.
pub fn cds_api_key() -> Result<String> {
    if let Some(k) = std::env::var_os("CDSAPI_KEY") {
        let k = k.to_string_lossy().trim().to_string();
        if !k.is_empty() {
            return Ok(k);
        }
    }
    if let Some(home) = std::env::var_os("HOME") {
        let rc = PathBuf::from(home).join(".cdsapirc");
        if let Some(k) = read_cdsapirc_key(&rc)? {
            return Ok(k);
        }
    }
    Err(Error::Transport {
        url: "cds://".to_string(),
        detail:
            "CDS API key not found: set CDSAPI_KEY or create ~/.cdsapirc with 'key: <your-key>'"
                .to_string(),
    })
}

/// A ready-to-register `cds`-realm [`AuthResolver`] carrying the `PRIVATE-TOKEN`
/// header, with the key resolved from `$CDSAPI_KEY` / `~/.cdsapirc`. Register it
/// with `Cache::builder().register_auth(Arc::new(cds_auth()?))`.
pub fn cds_auth() -> Result<StaticHeaderAuth> {
    Ok(StaticHeaderAuth::header(
        CDS_REALM,
        "PRIVATE-TOKEN",
        cds_api_key()?,
    ))
}

/// Read + parse the `key:` line from a `.cdsapirc` file. Absent file ⇒ `None`.
fn read_cdsapirc_key(path: &Path) -> Result<Option<String>> {
    match std::fs::read_to_string(path) {
        Ok(contents) => Ok(parse_cdsapirc_key(&contents)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(Error::io(Some(path.to_path_buf()), e)),
    }
}

/// Pull the first `key:` value out of a `.cdsapirc` body. Pure + testable.
fn parse_cdsapirc_key(contents: &str) -> Option<String> {
    for line in contents.lines() {
        if let Some(rest) = line.trim_start().strip_prefix("key:") {
            let key = rest.trim();
            if !key.is_empty() {
                return Some(key.to_string());
            }
        }
    }
    None
}

/// Turn an optional resolver into the header map applied to submit + poll. An
/// invalid header name/value is a clean error rather than a silently dropped
/// credential. No resolver ⇒ empty headers (the CDS server then returns 401).
fn auth_headers(url: &str, auth: Option<&dyn AuthResolver>) -> Result<HeaderMap> {
    let mut headers = HeaderMap::new();
    if let Some(resolver) = auth {
        for (name, value) in resolver.headers() {
            match (
                HeaderName::from_bytes(name.as_bytes()),
                HeaderValue::from_str(&value),
            ) {
                (Ok(n), Ok(v)) => {
                    headers.insert(n, v);
                }
                _ => {
                    return Err(transport_err(url, format!("invalid auth header '{name}'")));
                }
            }
        }
    }
    Ok(headers)
}

/// A [`Error::Transport`] for `url` with `detail`.
fn transport_err(url: &str, detail: String) -> Error {
    Error::Transport {
        url: url.to_string(),
        detail,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scheme_is_cds() {
        assert_eq!(CdsTransport::new().schemes(), &["cds"]);
    }

    #[test]
    fn base_url_trailing_slash_trimmed() {
        let t = CdsTransport::with_base_url("https://example.test/api/");
        assert_eq!(t.base_url, "https://example.test/api");
    }

    #[test]
    fn url_round_trips() {
        let req = r#"{"area":[50,-130,20,-60],"variable":["temperature"]}"#;
        let url = build_cds_url("reanalysis-era5-pressure-levels", req);
        assert_eq!(url, format!("cds://reanalysis-era5-pressure-levels?{req}"));
        let (dataset, request_json) = parse_cds_url(&url).unwrap();
        assert_eq!(dataset, "reanalysis-era5-pressure-levels");
        assert_eq!(request_json, req);
    }

    #[test]
    fn parse_rejects_malformed_urls() {
        assert!(parse_cds_url("https://x/y").is_err()); // wrong scheme
        assert!(parse_cds_url("cds://dataset-only").is_err()); // no '?'
        assert!(parse_cds_url("cds://?{}").is_err()); // empty dataset
        assert!(parse_cds_url("cds://ds?").is_err()); // empty request
        assert!(parse_cds_url("cds://ds?not-json").is_err()); // not JSON
        assert!(parse_cds_url("cds://ds?[1,2]").is_err()); // JSON but not an object
        assert!(parse_cds_url("cds://ds?{}").is_ok()); // empty object is valid
    }

    #[test]
    fn parses_key_from_cdsapirc_body() {
        let body = "url: https://cds.climate.copernicus.eu/api\nkey: abc-123-def\n";
        assert_eq!(parse_cdsapirc_key(body).as_deref(), Some("abc-123-def"));
        // No surrounding space, first match wins, trailing whitespace trimmed.
        assert_eq!(parse_cdsapirc_key("key:xyz  ").as_deref(), Some("xyz"));
        assert_eq!(
            parse_cdsapirc_key("key: first\nkey: second").as_deref(),
            Some("first")
        );
        // No key line, or an empty value.
        assert_eq!(parse_cdsapirc_key("url: only\n"), None);
        assert_eq!(parse_cdsapirc_key("key:   \n"), None);
    }

    #[test]
    fn reads_key_from_cdsapirc_file_or_absent() {
        let dir = tempfile::tempdir().unwrap();
        let rc = dir.path().join(".cdsapirc");
        std::fs::write(&rc, "url: https://cds/api\nkey: file-key\n").unwrap();
        assert_eq!(read_cdsapirc_key(&rc).unwrap().as_deref(), Some("file-key"));
        // A missing file is `None`, not an error.
        assert_eq!(read_cdsapirc_key(&dir.path().join("nope")).unwrap(), None);
    }

    #[test]
    fn auth_headers_from_resolver_and_empty_without() {
        let resolver = StaticHeaderAuth::header(CDS_REALM, "PRIVATE-TOKEN", "tok");
        let h = auth_headers("cds://ds?{}", Some(&resolver)).unwrap();
        assert_eq!(h.get("PRIVATE-TOKEN").unwrap(), "tok");
        let none = auth_headers("cds://ds?{}", None).unwrap();
        assert!(none.is_empty());
    }
}
